azerothcore/azerothcore-wotlk
# feat(Docker): Clean up build and runtime

I can be contacted on discord as mynameismeat\#2486. Please feel free to reach out/tag me in a channel if you have any questions, comments, concerns.

## Goals
- Smaller docker containers
  - Less extraneous tooling
  - Merge various build targets into single target per component
  - Utilize multi-stage builds to ensure contents of containers are only runtime dependencies
- remove "prod" container distinction
- Enable users to get started with minimal effort
- wrap the worldserver in a way that safely runs `server shutdown` when the container is shutdown
- Allow for creation/management of an "admin account"
- Maintain compatibility with documentation and `acore-docker` within reason.
- Use docker's built-in caching for ccache management
- Remove the need (and the ability :p) from the user to access the console.

## Changes Proposed:
-  This is a large PR with many files changed.
- Files changed and description of changes involved
    - `apps/docker/Dockerfile`
        - Remove "prod" distinction
        - utilize multi-stage container builds
        - Split the `ac-dev-server` out into its' own dockerfile
    - `docker-compose.yml`
        - Remove "prod" distinction
    - `apps/docker/docker-cmd.ts`
        - Remove "prod" distinction
    - `apps/docker/client-data.sh`
        - Script to install client data
        - function extracted out of `apps/installer/includes/functions.sh`
    - `apps/docker/entrypoint.sh`
        - Starts azerothcore components
        - starts worldserver wrapper if desired
            - The wrapper script can create a user on startup and runs `server shutdown 5s 0` in the case of a SIGTERM
        - templates out config files
            - There was a few ways I wanted to do this, but to avoid spooky action at a distance it just templates the `.conf` in place.
    - `env/docker/etc/*.dockerdist`
        - Template out MySQL connection
        - disable logging to a file. Docker handles logging, logging to a file as well doesn't really do anything for us
    - `deps/boost/CMakeLists.txt`
        - Allow boost to be statically linked if `DOCKER` env var is defined.
        - Boost libraries (even non `-dev` versions) took a large amount of space in the final libraries
    - `apps/compiler/includes/functions.sh`
        - Revise for params that can include spaces
    - `apps/bash_shared_includes.sh`
        - Allow Deno to be option with `SKIP_INSTALL_DENO` env var

## Issues Addressed:

- I don't think it counts if I created the issue, but #14644
- There's a wrapper that creates an admin user, similar outcome to #14774 but not exactly the same
- Can replace #14914

## SOURCE:


## Tests Performed:

- Able to build, login, and play without issues. Modules seem to work without issues. 


## How to Test the Changes:


0. Destroy previous tooling with `docker compose --profile app --profile prod down -v --remove-orphans`
1. Start AzerothCore via command `docker compose --profile app up --build`
2. Observe new account being created in docker logs
3. Login with creds `admin` `admin` and confirm functionality
4. Run ctrl+c to stop the server. Quickly, check in game message logs showing the shutdown
5. Install a module (`git clone https://github.com/azerothcore/mod-transmog --depth 1 modules/mod-transmog`)
6. Rebuild with the command from `1`, it should be similar in speed to builds with ccache
7. Confirm module works

## Known Issues and TODO List:


- the mysql client needs to be added to the runtime container, and it adds more space is needed imo. It'd be nice if we can ditch it entirely since `libmysqlclient` is needed anyway
- wrapping the worldserver isn't ideal. It'd be fantastic if we could 
   - create a user off of worldserver environment variables natively
   - Safely shutdown the server on SIGTERM/SIGINT
- Console access isn't necessarily available to the user. I personally don't think this is an issue, but I can see how it'd get in the way. Might be worth setting up SOAP or RA access by default.
- A health check would be great. Since I'm already using `expect` for wrapping the worldserver, it might not be too much of a stretch for it to access telnet and run `server info` or something


## How to Test AzerothCore PRs
 
When a PR is ready to be tested, it will be marked as **[WAITING TO BE TESTED]**.

You can help by testing PRs and writing your feedback here on the PR's page on GitHub. Follow the instructions here:

http://www.azerothcore.org/wiki/How-to-test-a-PR

**REMEMBER**: when testing a PR that changes something **generic** (i.e. a part of code that handles more than one specific thing), the tester should not only check that the PR does its job (e.g. fixing spell XXX) but **especially** check that the PR does not cause any regression (i.e. introducing new bugs).

**For example**: if a PR fixes spell X by changing a part of code that handles spells X, Y, and Z, we should not only test X, but **we should test Y and Z as well**.
